### PR TITLE
Fix casting of scaled_float in sorts

### PR DESCRIPTION
**Pull request by nik9000**
Previously we'd get a `ClassCastException` when you tried to use
`numeric_type` on `scaled_float`. Oops! This cleans up the CCE and moves
some code around so the casting actually works.

### CODE REVIEW

**Comment by nik9000** [comment]
It'd be super reasonable for you to ask "why in the world would you want to cast a `scale_float` to `long` or a `date` or `date_millis` or whatever?" Its a great question! I believe this feature was added so you could sort results from two indices with different mappings. Folk often make indices for some time period and when they change the mapping of a field they don't re-index their old data because it'd take forever. This lets you merge those fields together for sorting. It comes up. Truly. Maybe doing it with a `scaled_float` doesn't come up often though!

**Comment by nik9000** [comment]
@elasticmachine, run elasticsearch-ci/2

**Comment by javanna** [comment]
this makes sense to me but I think we should find a reviewer who's more familiar with mappings code than me :) maybe @romseygeek ?

**Comment by nik9000** [comment]
So pinged!

**Comment by romseygeek** [comment]
This isn't a bit of mappings that I know a lot about, alas - somebody who knows more about aggs than me is probably a better bet...

File: server/src/main/java/org/elasticsearch/index/fielddata/plain/SortedNumericIndexFieldData.java [hunk 1]
```diff
@@ -113,95 +101,29 @@ public final Index index() {
...
-     * Builds a {@linkplain BucketedSort} for the {@code targetNumericType},
-     * casting the values if their native type doesn't match.
-     */
-    public BucketedSort newBucketedSort(NumericType targetNumericType, BigArrays bigArrays, @Nullable Object missingValue,
-            MultiValueMode sortMode, Nested nested, SortOrder sortOrder, DocValueFormat format,
-            int bucketSize, BucketedSort.ExtraData extra) {
-        return comparatorSource(targetNumericType, missingValue, sortMode, nested)
-                .newBucketedSort(bigArrays, sortOrder, format, bucketSize, extra);
+    protected boolean sortRequiresCustomComparator() {
+        return numericType == NumericType.HALF_FLOAT;
```
Comment (polyfractal) [inline server/src/main/java/org/elasticsearch/index/fielddata/plain/SortedNumericIndexFieldData.java:106]: Maybe leave a comment in here about why half_floats need a custom comparator?  I'm assuming their encoding makes them non-ordered or something?

**Review by polyfractal — APPROVED** [review]
Left a comment but otherwise LGTM.  I'm far from an expert in this area of the code either, so grain of salt :)

File: modules/mapper-extras/src/test/resources/rest-api-spec/test/scaled_float/10_basic.yml [hunk 2]
```diff
@@ -97,8 +97,28 @@ setup:
 
   - do:
       search:
-        rest_total_hits_as_int: true
```
Comment (jimczi) [inline modules/mapper-extras/src/test/resources/rest-api-spec/test/scaled_float/10_basic.yml:100]: thanks

**Review by jimczi — APPROVED** [review]
LGTM, nice cleanup

File: server/src/main/java/org/elasticsearch/index/fielddata/plain/SortedNumericIndexFieldData.java (hunk 1 above)
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/index/fielddata/plain/SortedNumericIndexFieldData.java:106]: I believe that is so. I'll dig and leave a comment.

**Pull request merged**